juju/model.py

```diff
@@ -3212,6 +3245,222 @@ def _raise_for_status(entities: dict[str, list[str]], status: Any):
+                )
+
+        for unit_name, unit in units.items():
+            assert unit.workload_status
+            if unit.workload_status.status == "error" and raise_on_error:
+                raise JujuUnitError(
+                    f"{unit_name!r} workload has errored: {unit.workload_status.info!r}"
+                )
+
+        for app_name in apps:
+            app = full_status.applications[app_name]
+            assert isinstance(app, ApplicationStatus)
+            assert app.status
+            if app.status.status == "error" and raise_on_error:
+                raise JujuAppError(f"{app_name!r} has errored: {app.status.info!r}")
+
+        for unit_name, unit in units.items():
+            assert unit.workload_status
+            if unit.workload_status.status == "blocked" and raise_on_blocked:
+                raise JujuUnitError(
+                    f"{unit_name!r} workload is blocked: {unit.workload_status.info!r}"
+                )
+
+        for app_name in apps:
+            app = full_status.applications[app_name]
+            assert isinstance(app, ApplicationStatus)
+            assert app.status
+            if app.status.status == "blocked" and raise_on_blocked:
+                raise JujuAppError(f"{app_name!r} is blocked: {app.status.info!r}")
+
+        for unit_name, unit in units.items():
+            assert unit.agent_status
+            idle_times.setdefault(unit_name, now)
+            if unit.agent_status.status != "idle":
+                idle_times[unit_name] = now
+
+        for app_name in apps:
+            ready_units = []
+            app = full_status.applications[app_name]
+            assert isinstance(app, ApplicationStatus)
+            for unit in _app_units(full_status, app_name).values():
+                assert unit.agent_status
+                assert unit.workload_status
+
+                if unit.agent_status.status != "idle":
+                    continue
+                if status and unit.workload_status.status != status:
+                    continue
+
+                ready_units.append(unit)
+
+            if wait_for_exact_units is None and len(ready_units) < _wait_for_units:
+                logger.info(
+                    "Waiting for app %r units %s/%s",
+                    app_name,
+                    len(ready_units),
+                    _wait_for_units,
+                )
+                return False
+
+            if (
+                wait_for_exact_units is not None
+                and len(ready_units) != wait_for_exact_units
+            ):
+                logger.info(
+                    "Waiting for app %r units %s/%s",
+                    app_name,
+                    len(ready_units),
+                    _wait_for_units,
+                )
+                return False
+
+        if busy := [n for n, t in idle_times.items() if expected_idle_since < t]:
+            logger.info("Waiting for %s to be idle enough", busy)
+            return False
```

Comment:
Suggestions:
- rename `idle_times` to `last_active` or `last_non_idle` (more obvious what we're tracking here)
- could we use a property or something to exfiltrate this state between calls instead of a mutable argument? e.g. `self._last_active_unit_times` (any concurrency concerns with this?)
- I like the one liner, but to understand it I mentally expand to more verbose and explicit names, (and swap conditional direction)
```py
busy_units = [
    unit_name
    for unit_name, last_active_time
    in last_active.items()
    if last_active_time > required_idle_since
]
if busy_units: ...
```

Reply:
Let me think about that.
last active time is not actually tracked.
What is tracked is "unit X was continuously idle since sampled at 11:22:33", for some degree of "continuously" obv.

Reply:
Ah yep, so `last_active_time` is not really an accurate name. `idle_since` maybe

Reply:
done